docs/cardano-monads.rst

```diff
@@ -0,0 +1,467 @@
+  could mean hours of compilation and required up to 65 GIGABYTES of RAM
+  (ridiculous!). This was because GHC generated an exponential amount of
+  coercions (as evidenced by investigating .hi-files). Basically, @int-index no longer
+  can recommend Ether to people as there is no good solution to this known.
+
+Conclusion:
+
+* Migration to Ether allowed us to remove an immense amount of boilerplate,
+  modularize the code, and get good run-time performance. However, lack of
+  run-time configuability was quite inconvenient, and bad compile-time
+  performance marked this approach a no-go.
+
+The Modern Era
+---------------
+
+After we've realized what led to bad compile-time performance, @int-index came up with an
+idea of ``ExecMode``. Basically, we continued to use classes from ``Ether``, but
+rather than having numerous ``IdentityT`` layers there was a single ``newtype``
+wrapper around ``ReaderT ModeEnv Production`` at the bottom. This solved the
+compile-time performance issue completely at the cost of a moderate increase in
+boilerplate.
+
+However, FPComplete began to see ``ether`` as an Enemy of The State, and asked
+us to purge its remains. Now we were supposed to remove all our custom classes,
+replacing them with method records. Those records would go into a ``ReaderT``
+and be passed everywhere manually (as opposed to instance search). Instead of
+distinct ``Ether.MonadReader`` constraints, now we had to use ``MonadReader
+ctx``, passing an annoying ``ctx`` parameter everywhere, and placing constraints
+on it. The final transformer stack is just ``ReaderT ModeCtx Production``, and
+not even a newtype on top.
+
+Technically, we're still in the process of migration, as we haven't removed all
+of our custom classes yet. Just to clarify: in this section we'll discuss the
+current transitional state, and it's more painful than what was actually
+proposed by FPComplete.
+
+
+Also, there is compliсation, that we actually have three aproaches combined:
+	
+* General approach a la ExecMode(s)
+* Usage of reflection (which is in fact custom case for ReaderT for constant data within single execution)
+* Method dictionaries (for supporting SendActions, a primitive from networking)
+* ReaderT with method dictionaries
+
+General approach
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Now our code follows this pattern::
+
+    -- effect definitions
+
+    class MonadBaz
+        baz :: ...
+
+    defaultBaz = ...
+
+    class MonadZaz
+        zaz :: ...
+
+    defaultZaz = ...
+
+
+    -- mode definitions
+
+    data QuuxCtx = ...
+
+    type QuuxMode = ReaderT QuuxCtx Production
+
+    instance HasFoo ModeCtx
+    instance HasBar ModeCtx
+
+    instance MonadBaz QuuxMode
+        baz = defaultBaz
+
+    instance MonadZaz QuuxMode
+        zaz = defaultZaz
+
+
+Pros:
+
+* Good compile-time performance. There's only a single layer of transformers,
+  instance search is quick, no coercions involved.
+
+Cons:
+
+* Bad run-time configurability. (Same as above)
```

Comment:
Can you elaborate? As an example i can provide `DBSum` which is pure or real db. And it works nicely with `ReaderT` approach, you can choose one option it in the runtime.

I'm not trying to claim it is optimal (it's far from that, we have unsafe lenses to deal with the fact one can switch db after it's set for the first time), but I'd maybe like to see more details on what problems do we have.

Reply:
I believe I addressed it:

> As discovered later, this can be addressed with closed effect sums, but then we lose extensibility, so it's still an issue.

The problem is that you need to know about all possible implementations up front.

Reply:
How it is a problem at all? I know that there will be two implementations of `DB`.
The problem is that having the sum doesn't restrict the context to change the implementation in the runtime, what we actually would like to ban.

Reply:
> How it is a problem at all?

It's a problem if you care about modular code where effects and implementations are separated.

Reply:
I seem to miss the point, can you elaborate again using the `DBSum` example? What is the problem you're trying to emphasize? 
inb4 `cardano-sl` is not a "library" in a common sense -- whenever we want to add extra implementation, we can add it to library directly (if it was a no-change-only-import library, it would be a problem to extend these types).

Reply:
There's now a section on effect context sums.